### ISSUE TITLE
How to fetch token via API?

### ISSUE THREAD

**Issue by KakaWangSAP** [kyma-project/kyma#6090]
Hi,
Here is the background, I have a kyma console UI, and I created a Lambda, and trigger it via https, and secured it. and save it. The token generated like below:
![image](https://user-images.githubusercontent.com/51769984/67375061-a9036080-f5b4-11e9-883d-a6683a9dd107.png)
Also I can get token by click "Fetch token" here:
![image](https://user-images.githubusercontent.com/51769984/67375228-e8ca4800-f5b4-11e9-9e67-df6ce5f6c697.png)
My question is: Is there any token api than I can invoke to fetch token just like above kyma console UI does?

**Comment by abbi-gaurav** [comment]
@kubadz, could you please add some details about why this is a question?

**Comment by piotrmsc** [comment]
@abbi-gaurav Because it is a question? "My question is: Is there any token api than I can invoke to fetch token just like above kyma console UI does?"

@KakaWangSAP Thanks for testing kyma!  There's no such thing as token API in the kyma. The token you see in the console is your personal ID Token fetch by kyma console in the authentication process.

You can fetch the token by yourself as well by doing implicit or authorization code grant flow againts dex. This is standard Oauth2 flow, please have in mind that implicit flow is desired mostly for web applications, where client_secret of oauth2 client should not be publicly available. Authorization code grant requires a backend in which code exchange should happen.
Currently, in kyma we do not have client management in dex, meaning you will need to manually add a new client to dex configuration. Dex config is here https://github.com/kyma-project/kyma/blob/master/resources/dex/templates/dex-config-map.yaml. TO add a client you can add an override and trigger kyma update or (with timeout) edit the configmap of dex and restart the pod.

But if you are intrested in exposing an API with tokens without user context (machine to machine communications) you can tests our new API Gateway and it's oauth2 integration described here https://kyma-project.io/docs/components/api-gateway-v2/#overview-overview

**Comment by KakaWangSAP** [comment]
@piotrmsc Thanks for you replay, I definitely want to go for the second way API Gateway v2, but after went through the whole overview page, all things are done by commend line which is very hard to follow, Is it also can be done in console UI? Another question is, are you plan to have an easy to use token API in Kyma further version?

**Comment by MingZhang-PS** [comment]
Hi @piotrmsc
I am following the guideline at  https://kyma-project.io/docs/components/api-gateway-v2/#overview-overview
Now the oauth2client CR has been created successfully
![image](https://user-images.githubusercontent.com/20200532/67661552-7da9b880-f99c-11e9-8f5b-a0901290ab28.png)
However, when I tried to request the token, it gives me error
![image](https://user-images.githubusercontent.com/20200532/67661704-d8dbab00-f99c-11e9-9225-10dabc85fe43.png)
So how can I know if oauth2 client has been created successfully via the ORY Hydra's API? Or if there is some other possible reason for this error?

**Comment by piotrmsc** [comment]
@KakaWangSAP there will not be anything like token api, everything is standarized via oauth2 and OIDC. In the UI there will be new UI to expose API via API Gateway and client registration UI as well. 

And because you have custom resource representing client you can automate that in your code :)

**Comment by piotrmsc** [comment]
@MingZhang-PS Hi,

There is some time needed to propagate the information about new client (from custom resource) to hydra, does it still fail with 401? Error indicates the client was not registered in hydra

**Comment by piotrmsc** [comment]
btw, I encourage you all to join our public slack ;-) I think we might get quicker to you there ;-)

**Comment by MingZhang-PS** [comment]
Hi @piotrmsc 
I tried to get some log from Hydra Maester, it shows some error like below. 
But how can I make the user "system:serviceaccount:kyma-system:hydra-maester-account" to be able to create resource "secrets"?

**2019-10-28T07:36:39.095Z	ERROR	controllers.OAuth2Client	error creating secret for client kaka-pass-2/kaka 	{"oauth2client": "register", "error": "secrets is forbidden: User "system:serviceaccount:kyma-system:hydra-maester-account" cannot create resource "secrets" in API group "" in the namespace "kaka""}**
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/ory/hydra-maester/controllers.(*OAuth2ClientReconciler).registerOAuth2Client
	/go/src/github.com/ory/hydra-maester/controllers/oauth2client_controller.go:123
github.com/ory/hydra-maester/controllers.(*OAuth2ClientReconciler).Reconcile
	/go/src/github.com/ory/hydra-maester/controllers/oauth2client_controller.go:86
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:216
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:192
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:171

**Comment by MingZhang-PS** [comment]
Talked more with Piotr. The root cause is clear, please **make sure the kyma version to be >=1.6 when you start to create oauth2 client by Hydra.** 
And if you are in lower version, such as kyma 1.5, you can workaround by following commands
```
cat <<EOF | kubectl apply -f -
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: hydra-maester-secrets-role-global
rules:
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "watch", "create"]
```
```
cat <<EOF | kubectl apply -f -
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: hydra-maester-role-binding-global
subjects:
  - kind: ServiceAccount
    name: hydra-maester-account # Service account assigned to the controller pod.
    namespace:  kaka
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: hydra-maester-role-global
```
After that, you can follow https://kyma-project.io/docs/components/api-gateway-v2/#tutorials-expose-and-secure-a-service to create oauth2 client and get access token without trouble

**Comment by MingZhang-PS** [comment]
@KakaWangSAP  you are issue submitter, please close this issue :-)
